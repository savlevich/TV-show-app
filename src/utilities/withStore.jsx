import React from 'react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';

const withStore = (Component, initialState = {}) => () => {
  const mockStore = configureStore();
  const store = mockStore(initialState);
  return <Provider store={store}>{Component}</Provider>;
};

export default withStore;
