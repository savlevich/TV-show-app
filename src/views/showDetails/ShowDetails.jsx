import React, { useEffect } from "react";
import { Link } from 'react-router-dom';
import { useDispatch, useSelector } from 'react-redux';
import { fetchShowDetailsData, fetchShowEpisodesData } from './ducks/actions';
import {
  getShowNameSelector,
  getShowDescriptionSelector,
  getShowIdSelector,
  getShowEpisodesSelector,
  getShowPictureSelector
} from './ducks/selectors';
import ContentWrapper from '../../components/ContentWrapper';
import './styles.scss';

const ShowDetails = () => {
  const dispatch = useDispatch();
  const showId = useSelector(getShowIdSelector);
  const showName = useSelector(getShowNameSelector);
  const pictureUrl = useSelector(getShowPictureSelector);
  const showEpisodes = useSelector(getShowEpisodesSelector);
  const showDescription = useSelector(getShowDescriptionSelector);

  useEffect(() => {
    const fetchData = async () => {
      dispatch(fetchShowDetailsData());
    };
    fetchData()
  }, [dispatch]);

  useEffect(() => {
    const fetchData = async showId => {
      dispatch(fetchShowEpisodesData(showId));
    };

    if (showId) {
      fetchData(showId);
    }
  }, [dispatch, showId]);

  return (
    <ContentWrapper className="show">
      <div className="show__info">
        <h2 className="show__title">{showName}</h2>
        <div className="show__picture__wrapper">
          <img width="300" height="400" className="show__picture" alt="show" src={pictureUrl} />
        </div>
        <div className="show__desc" dangerouslySetInnerHTML={{ __html: showDescription }} />
      </div>
      <div className="show__episodes_title">Episodes</div>
      <ul className="show__episodes_list">
        {showEpisodes && showEpisodes.map(({ name, id, season, number }) =>
          <li className="show__episodes_list_item" key={name}>
            <Link className="show__episodes_list_item" to={`/episodes/${id}/the-powerpuff-girls`}>
              Season {season} episode {number} {name}
            </Link>
          </li>
        )}
      </ul>
    </ContentWrapper>
  )
};

export default ShowDetails;