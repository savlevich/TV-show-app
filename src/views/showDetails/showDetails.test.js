import React from 'react';
import { render } from '@testing-library/react';
import { useParams, MemoryRouter } from 'react-router-dom';
import withStore from '../../utilities/withStore';
import ShowDetails from './ShowDetails';

const store = {
  show: {
    showDetails: {
      name: 'The Powerpuff Girls',
      id: 'test'
    },
    showEpisodes: [
      { name: 'Escape from Monster Island'},
    ]
  },
};

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useParams: jest.fn(),
}));

describe('ShowDetails', () => {
  test('should render ShowDetails', () => {
    useParams.mockReturnValue({ episodeId: 'test' });

    const ElementWithStore = withStore(
      <MemoryRouter>
        <ShowDetails />
      </MemoryRouter>, store);
    const { asFragment } = render(<ElementWithStore />);

    expect(asFragment()).toMatchSnapshot();
  });
});
