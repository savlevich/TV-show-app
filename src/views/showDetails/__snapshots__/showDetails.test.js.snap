// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShowDetails should render ShowDetails 1`] = `
<DocumentFragment>
  <div
    class="content-wrapper"
  >
    <div
      class="show__info"
    >
      <h2
        class="show__title"
      >
        The Powerpuff Girls
      </h2>
      <div
        class="show__picture__wrapper"
      >
        <img
          alt="show"
          class="show__picture"
          height="400"
          src=""
          width="300"
        />
      </div>
      <div
        class="show__desc"
      />
    </div>
    <div
      class="show__episodes_title"
    >
      Episodes
    </div>
    <ul
      class="show__episodes_list"
    >
      <li
        class="show__episodes_list_item"
      >
        <a
          class="show__episodes_list_item"
          href="/episodes/undefined/the-powerpuff-girls"
        >
          Season  episode  Escape from Monster Island
        </a>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;
